config/src/config/network_config.rs

```diff
@@ -39,6 +39,8 @@ pub struct NetworkConfig {
     // Otherwise, any node can connect. If this flag is set to true, `enable_noise` must
     // also be set to true.
     pub enable_remote_authentication: bool,
+    // Enable this network to use either gossip discovery or onchain discovery.
+    pub discovery_method: DiscoveryMethod,
```

Comment:
This feels interesting. My gut feeling is we should not have only one method for discovery:
1) OnChain is always available and we should just use that if we get it
2) We should have to manually turn on Gossip

So I'd prefer we just have a flag that explicitly says enable_gossip_discovery

Reply:
enumerating the possible network instances:

1. validator network: probably always want onchain discovery (but just validator info)
2. validator full node <-> validator: i don't think (?) we need any discovery here
3. validator full node <-> other validator full node (for reliability, see https://fb.quip.com/IgOaA4qmHTId): onchain discovery (but just full node info)
4. public full node / client <-> validator full node: public full node uses onchain discovery (but just full node info). validator full node isn't using any discovery in this network instance, since it's a "permission-less" network.
5. node-internal communication: no discovery here either

so in the end state, there's 3 cases:
1. no discovery
2. onchain discovery (validator info only)
3. onchain discovery (fullnode info only)

however, until we get genesis config code more fleshed out and working with cluster-test, devnet, and pre-mainnet I don't feel comfortable enough turning on onchain discovery, which is why this diff just has gossip discovery as the default.

maybe that's a good enough reason to spend some time add support for multiple discovery sources to connectivity manager so we can support gossip discovery as a connectivity hint in cases where the config tooling isn't ready for onchain discovery.